### PR TITLE
refactor: extract reusable logic from Agent Writer to HTTP Writer

### PR DESCRIPTION
**Pull request by jirikuncar**
This is preparing ground for CI Visibility Agentless support that will reuse retry logic from Agent Writer. Also we would like to keep the possibility for metrics calculation and submission in background.

https://github.com/DataDog/dd-trace-py/pull/3654#discussion_r866849329

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: releasenotes/notes/HTTP-Writer-7bd3c1bd0234cd2d.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+other:
+  - |
+    Refactors the Agent Writer to use the new HTTP Writer interface.
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/HTTP-Writer-7bd3c1bd0234cd2d.yaml:4]:
```suggestion
```

this is an internal refactor

File: ddtrace/internal/writer.py [hunk 2]
```diff
@@ -298,7 +285,7 @@ def __init__(
                 }
             )
 
-        self._encoder = Encoder(
+        self._encoder = encoder_cls(
             max_size=self._buffer_size,
```
Comment (majorgreys) [inline ddtrace/internal/writer.py:289]: We discussed moving these encoder parameters out of the writer in a separate refactor.

File: ddtrace/internal/writer.py [hunk 3]
```diff
@@ -363,22 +350,6 @@ def _set_keep_rate(self, trace):
         if trace:
             trace[0].set_metric(KEEP_SPANS_RATE_KEY, 1.0 - self._drop_sma.get())
```
Comment (majorgreys) [inline ddtrace/internal/writer.py:352]: This should be specified so subclasses need to implement.

File: ddtrace/internal/writer.py [hunk 4]
```diff
@@ -272,23 +271,11 @@ def __init__(
             "Datadog-Meta-Lang-Version": compat.PYTHON_VERSION,
             "Datadog-Meta-Lang-Interpreter": compat.PYTHON_INTERPRETER,
             "Datadog-Meta-Tracer-Version": ddtrace.__version__,
-            "Datadog-Client-Computed-Top-Level": "yes",
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/writer.py:275]: Merge conflict?

File: ddtrace/internal/writer.py [hunk 5]
```diff
@@ -392,17 +363,17 @@ def _put(self, data, headers):
...
-                log.debug("creating new agent connection to %s with timeout %d", self.agent_url, self._timeout)
-                self._conn = get_connection(self.agent_url, self._timeout)
+                log.debug("creating new agent connection to %s with timeout %d", self.intake_url, self._timeout)
+                self._conn = get_connection(self.intake_url, self._timeout)
             try:
-                self._conn.request("PUT", self._endpoint, data, headers)
+                self._conn.request(self.REQUEST_HTTP_METHOD, self._endpoint, data, headers)
                 resp = compat.get_connection_response(self._conn)
                 t = sw.elapsed()
                 if t >= self.interval:
```
Comment (majorgreys) [inline ddtrace/internal/writer.py:372]: We should be emitting this timing as a metric as well.

File: releasenotes/notes/HTTP-Writer-7bd3c1bd0234cd2d.yaml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+---
```
Comment (majorgreys) [inline releasenotes/notes/HTTP-Writer-7bd3c1bd0234cd2d.yaml:1]: We can drop this release note as the refactor is only to internal api.

File: ddtrace/internal/writer.py [hunk 7]
```diff
@@ -225,20 +226,19 @@ def flush_queue(self):
...
     RETRY_ATTEMPTS = 3
 
+    REQUEST_HTTP_METHOD = "PUT"
+
     def __init__(
         self,
-        agent_url,  # type: str
+        intake_url,  # type: str
+        endpoint,  # type: str
+        # TODO (jirikuncar): define encoder protocol
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/writer.py:240]:
```suggestion
```

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]
I'm not a huge fan of this refactor as it is but I think it's a step in the right direction. There's too much coupling to the specific data model used by /v0.3/v0.4/v0.5 traces which we're just splitting into two classes. The parent class `HTTPWriter` still has dependence on the same logic and contains agent specific logic.

If reusability is the objective then I think a "has-a" relationship would be better than a "is-a". That is, encapsulate the transmission state and the corresponding metrics into its own class and have writers maintain an instance of that class.

This PR did separate a bit of the logic, I think we just need to pull out the last bits of agent-specific code and change the class relationship.

**Comment by emmettbutler** [comment]
@Kyle-Verhoog I'm working on a fresh attempt at this change [here](https://github.com/DataDog/dd-trace-py/compare/1.x...emmett.butler/civ-agentless), and I'm not sure I can see the need for the "has-a" relationship you described here. I think I've got a change that pretty well separates the concerns that are AgentWriter-specific from those that apply to the HTTPWriter using an "is-a" relationship. I'm not asking you to review that change, but could you provide any more detail on what "transmission state and corresponding metrics" would be better encapsulated in an instance attribute than a subclass relationship? Thanks a lot for loading this old comment back into your brain.

**Comment by Kyle-Verhoog** [comment]
@emmettbutler I think where I was coming from is the leaking of concerns between the base and subclasses. By doing an "is-a" relationship it is very easy to move a method up a layer of the classes and have the concerns leaked to the base. 

I think "is-a" can work it's just much easier to fall into implicitly coupling the classes because it's so easy to share state between them.

**Pull request closed without merging**